en/lessons/ecto/changesets.md

```diff
@@ -0,0 +1,372 @@
+
+```elixir
+@prohibited_words ["bad", "offensive", "suggestive"]
+
+def validate_name_appropriate(changeset) do
+  name_unsafe =
+    changeset
+    |> get_field(:name)
+    |> String.downcase()
+    |> String.split(" ", trim: true)
+    |> Enum.any?(name_parts, &(&1 in @prohibited_words))
+
+  if name_unsafe do
+    add_error(changeset, :name, "contains prohibited words")
+  else
+    changeset
+  end
+end
+
+```
+
+Above we introduced two new functions: [`get_field/3`](https://hexdocs.pm/ecto/Ecto.Changeset.html#get_field/3) and [`add_error/4`](https://hexdocs.pm/ecto/Ecto.Changeset.html#add_error/4).
+While we can probably guess what they're doing, let's walk through our validation anyways:
+
++ First we use `get_field/3` to retreive the value for `:name`.
+It's important to know that `get_field/3` will try to get the value from the changes first before falling back to the original data.
++ Next we downcase our string to make it easier to compare with our prohibited words.
++ We split it into parts (`"test user"` becomes `["test", "user"]`).
++ With `Enum.any?/2` we check if any part of our name is included in the prohibited words list.
++ Finally, if there we do have an unsafe name we rely on `add_error/3` to add a new error to our change.
+Here we pass in our changeset, the field to which the occur refers, and finally our error message.
+
+It is a good practice to always return a `%Ecto.Changeset{}` so we can use it in our `changeset/2` pipeline:
+
+```elixir
+def changeset(struct, params) do
+  struct
+  |> cast(params, [:age, :name])
+  |> validate_required([:age, :name])
+  |> validate_length(:name, min: 2)
+  |> validate_number(:age, greater_than: 0, less_than: 120)
+  |> validate_name_appropriate()
+end
+```
+
+Now we can give it a try:
+
+```elixir
+iex> Person.changeset(%Person{}, %{"name" => "Offensive Name"})
+#Ecto.Changeset<
+  action: nil,
+  changes: %{},
+  errors: [name: {"contains prohibited words", []}],
+  data: #Person<>,
+  valid?: false
+>
+```
+
+No surprise to us, our new validation rule works!
+
+## Adding changes programatically
+
+Sometimes you want to introduce changes to a changeset manually.
+We previously saw how to use `changes/2` to apply a map of changes to our changeset, now we'll look at a different approach using `put_change/3`.
+
+To demonstrate `put_change/3` we'll update our `changeset/2` function to allow `:name` to be blank and when it's blank set it to `"Anonymous"`.
+In order to accomplish this let us create a new `set_anonymous/1` function that accepts and returns a changeset as we did earlier:
+
+```elixir
+def set_anonymous(changeset) do
+  name = get_field(changeset, :name)
+
+  if is_nil(name) or name == "" do
+    put_change(changeset, :name, "Anonymous")
+  else
+    changeset
+  end
+end
+```
+
+Nothing to difficult, right?  If the `:name` is blank or an empty string we want to put a change on to the changeset making `:name` equal to `"Anonymous"`.
```

Comment:
Nothing too difficult, right?  If the `:name` is blank or an empty string, we want to put a change on to the changeset making `:name` equal to `"Anonymous"`.

Reply:
Just some food for thought:

I cannot find the exact article / research around this at the moment, but I remember reading some advice on writing tutorials that suggested avoiding expressions like "Nothing too difficult, right?", "This is super easy", "Few basic / simple steps" etc. Basically implying that something is very easy or simple. The thinking behind this is:

1) Readers who struggle with the lesson, for whatever reason, will find this kind of expressions discouraging (e.g. "Why do I struggle if this is supposed to be easy?")

2) Readers who do well might will feel lower sense of achievement.

Since then I stopped using those when writing tutorials / documentation / etc.

Again, this is just some food for thought 😃 I tend to over-think things sometimes 😄

Reply:
That makes perfect sense @svileng, I think so much so we that add a note to our styleguide about it.

It may be a day or two before I can get back to these updates but when I do I will fix this 😁